src/main/scala/linting/rule/LintAmbiguousModuleNames.scala

```diff
@@ -0,0 +1,278 @@
+  def getName(desiredName: String)(module: Module): String = {
+    val noNameModule = NamingStrategy.removeModuleNames(module)
+    NamingStrategy.appendHashCode(desiredName, "c", noNameModule.hashCode)
+  }
+}
+
+// use module content hash naming strategy
+case object ContentNamingStrategy extends NamingStrategy {
+  def getName(desiredName: String)(module: Module): String = {
+    val noInfoModule = NamingStrategy.removeModuleInfo(module).copy(name = NamingStrategy.emptyName)
+    NamingStrategy.appendHashCode(desiredName, "C", noInfoModule.hashCode)
+  }
+}
+
+object NamingStrategy {
+
+  final val emptyName: String = ""
+
+  def appendHashCode(desiredName: String, hashPrefix: String, hashCode: Int): String = {
+    s"${desiredName}_$hashPrefix" + f"${hashCode}%08X"
+  }
+
+
+  // remove name helpers
+
+  def removeTypeNames(tpe: Type): Type = tpe match {
+    case g: GroundType => g
+    case v: VectorType => v
+    case b: BundleType => b.copy(fields = b.fields.map { field =>
+      field.copy(name = emptyName)
+    })
+  }
+
+  def removeStatementNames(stmt: Statement): Statement = stmt match {
+    case i: DefInstance => i.copy(
+      name = emptyName,
+      module = emptyName,
+      info = NoInfo)
+    case i: WDefInstance => i.copy(
+      name = emptyName,
+      module = emptyName,
+      tpe = removeTypeNames(i.tpe),
+      info = NoInfo)
+    case _ => stmt
+      .mapStmt(removeStatementNames)
+      .mapString(_ => emptyName)
+      .mapInfo(_ => NoInfo)
+      .mapType(removeTypeNames)
+  }
+
+  def removePortNames(port: Port): Port = {
+    port.copy(name = emptyName, tpe = removeTypeNames(port.tpe), info = NoInfo)
+  }
+
+  def removeModuleNames(mod: Module): Module = {
+    mod.copy(
+      ports = mod.ports.map(removePortNames(_)),
+      body = removeStatementNames(mod.body),
+      name = emptyName,
+      info = NoInfo)
+  }
+
+  // remove Info helpers
+
+  def removeStatementInfo(stmt: Statement): Statement = {
+    stmt.mapStmt(removeStatementInfo).mapInfo(_ => NoInfo)
+  }
+
+  def removePortInfo(port: Port): Port = {
+    port.copy(info = NoInfo)
+  }
+
+  def removeModuleInfo(mod: Module): Module = {
+    mod.copy(
+      ports = mod.ports.map(removePortInfo(_)),
+      body = removeStatementInfo(mod.body),
+      info = NoInfo)
+  }
+}
+
+
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+case class ModuleNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): ModuleNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+case object StabilizeNamesAspect extends Aspect[RawModule] {
+  def toAnnotation(top: RawModule): AnnotationSeq = {
+    Select.collectDeep(top) {
```

Comment:
For these cases, can we actually compute a better name using the parameters of the module? E.g. using the queue datatype for the queue name?

Reply:
i think at this point you should be throwing it back to the designer.
Do we have a clean API for designers to deal with these issues, other than setting .desiredName in the module's code itself? It would be nice to be able to do this "after the fact" for a design where you suddenly start getting collisions

Reply:
I was trying to demonstrate (b) with this aspect. it sets the name for all instances, but you could also annotate specific instances by matching on a parent module and dotting to the specific instances.

Reply:
I tried using the type name and the entries for the queue, but it looks like that isn't enough because of `pipe` and `flow` params. and those params aren't public

Reply:
side note: we should make those parameters public, IMO there is no good reason for them not to be

Reply:
Am I correct that your test cases in your PR description don't capture these changes for Queue and TLMonitor? Can you show an example of what it looks like when you apply this aspect?

Reply:
yeah I'll add more tests.

Reply:
> side note: we should make those parameters public, IMO there is no good reason for them not to be

https://github.com/freechipsproject/chisel3/pull/1452